design and build Project structure

yazan-hamamdi opened the pull request.

yazan-hamamdi committed:
design and build Project structure

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Enums/Class.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace AirportTicketBookingSystem.Enums
+{
+    public enum Class
```
Add explicit integer values (`Economy = 1`, `Business = 2`, `FirstClass = 3`) to prevent breaking changes when enum members are reordered, ensure consistent database storage and API serialization, and make debugging easier by having predictable numeric values. Also consider renaming from `Class` to `ServiceClass` or `TravelClass` to avoid potential conflicts with the C# class keyword.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Models/Passenger.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+        public string FullName { get; set; }
+        public string PassportNumber { get; set; }
+        public List<Booking> Booking { get; private set; } = new();
```
`Bookings`

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Models/Passenger.cs:
```diff
@@ -0,0 +1,19 @@
+﻿
+namespace AirportTicketBookingSystem.Models
```
you can use file scoped namespaces to reduce indentation: 
```csharp
namespace AirportTicketBookingSystem.Models;
```

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Models/Flight.cs:
```diff
@@ -0,0 +1,24 @@
+        public string ArrivalAirport { get; set; }
+        public DateTime DepartureDate { get; set; }
+        public List<Booking> Bookings { get; private set; } = new();
+        public Flight(string departureCountry, string destinationCountry, string departureAirport,
```
new line before constructor

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Models/FirstClass.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public class FirstClass : ISeatClass
+    {
+        private const decimal multiplier = 4;
```
the name is not descriptive enough, multiplier used for what purpose exactly?

RandaZraik replied:
same comment for the other classes

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Models/FirstClass.cs:
```diff
@@ -0,0 +1,12 @@
+    {
+        private const decimal multiplier = 4;
+        public Class Name => Class.FirstClass;
+        public decimal CalculatePrice(decimal basePrice) => basePrice * multiplier;
```
new line before the method

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Models/Booking.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace AirportTicketBookingSystem.Models
+{
+    public class Booking
```
do you need 'id's for these models? to distinguish each record and treat it as a unique id.

yazan-hamamdi committed:
-Add explicit integer values for enum
-use file scoped namespaces
-Removing ambiguity for Magic numbers
-Improve formatting of files

yazan-hamamdi committed:
-Add Id's for models

yazan-hamamdi committed:
Adding a class to handle CSV files as a generic data store, ensuring it can work with all lists in the project

yazan-hamamdi committed:
implement repository pattern

yazan-hamamdi committed:
Add GetById method for all repos

yazan-hamamdi committed:
Make the Repositories CRUD only to achieve separation of concerns

yazan-hamamdi committed:
removed DeleteWhere method because it’s both a search and delete method, which could be confusing in the repository

yazan-hamamdi committed:
Improved repository names to be cleaner
Added basic services

yazan-hamamdi committed:
Add cascade operations for models that have composition relationship

yazan-hamamdi committed:
Improved the booking model to ensure the price can’t be altered and is calculated automatically

yazan-hamamdi committed:
improvement to fix the booking price issue

yazan-hamamdi committed:
Add a flight search method for the passenger
Search bookings for the manager
Methods for editing, deleting, and viewing bookings for the passenger

yazan-hamamdi committed:
Improved coordination between methods
Ensured that interfaces include all method signatures

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Enums/TravelClass.cs:
```diff
@@ -0,0 +1,11 @@
+
+    public enum TravelClass
+    {
+        none = 0,
```
PascalCase ```None```.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Models/Passenger.cs:
```diff
@@ -0,0 +1,23 @@
+﻿
+namespace AirportTicketBookingSystem.Models;
+
+    public class Passenger
```
please fix the indentations for the whole class

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Models/FirstClass.cs:
```diff
@@ -0,0 +1,21 @@
+
+    public class FirstClass : ISeatClass
+    {
+        private const decimal priceFactor = 4;
```
constants should be in PascalCase ```PriceeFactor```.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/IServices/IBookingService.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using AirportTicketBookingSystem.Models;
+
+namespace AirportTicketBookingSystem.IServices
```
the name of this folder ```IServices``` is inconvenient, move these interfaces to the ```Interfaces``` folder.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Interfaces/ISeatClass.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace AirportTicketBookingSystem.Interfaces;
+
+    public interface ISeatClass
```
this can just move to the models folder.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/IServices/IPassengerService.cs:
```diff
@@ -0,0 +1,33 @@
+        void UpdatePassenger(int id, Passenger newPassenger);
+        void DeletePassenger(int id);
+        List<Passenger> GetAllPassengers();
+        List<Flight> SearchAvailableFlights(
```
since this is mainly searching for flights and returning a list of flights, then this should belong to the ```FlightService```.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/IServices/IPassengerService.cs:
```diff
@@ -0,0 +1,33 @@
+            decimal? minPrice = null,
+            decimal? maxPrice = null
+        );
+        List<Booking> GetBookingsForPassenger(int passengerId);
```
this is returning bookings so should be moved to the ```BookingService```

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/IServices/IPassengerService.cs:
```diff
@@ -0,0 +1,33 @@
+            decimal? maxPrice = null
+        );
+        List<Booking> GetBookingsForPassenger(int passengerId);
+        void CancelBooking(int passengerId, int bookingId);
```
same thing, needs to be moved to the ```BookingService```

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/IServices/IPassengerService.cs:
```diff
@@ -0,0 +1,33 @@
+        );
+        List<Booking> GetBookingsForPassenger(int passengerId);
+        void CancelBooking(int passengerId, int bookingId);
+        void ModifyBooking(int passengerId, int bookingId, Booking newBooking);
```
same thing, needs to be moved to the ```BookingService```

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/IServices/IManagerService.cs:
```diff
@@ -0,0 +1,24 @@
+        void AddManager(Manager manager);
+        void UpdateManager(int id, Manager updatedManager);
+        void DeleteManager(int id);
+        List<Booking> FilterBookingsForManager(
```
this should be moved to the ```BookingService```.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/IRepositories/IBookingRepo.cs:
```diff
@@ -0,0 +1,13 @@
+﻿using AirportTicketBookingSystem.Models;
+
+namespace AirportTicketBookingSystem.IRepositories
```
move these interfaces to the ```Interfaces``` folder, maybe under ```Repositories``` folder there

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/IRepositories/IManagerRepo.cs:
```diff
@@ -0,0 +1,15 @@
+        void AddManager(Manager manager);
+        void UpdateManager(Func<Manager, bool> predicate, Manager newManager);
+        void DeleteManager(int id);
+
```
please remove extra newline

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/IRepositories/IManagerRepo.cs:
```diff
@@ -0,0 +1,15 @@
+        void DeleteManager(int id);
+
+    }
+
```
please remove extra newline

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/DTOs/SeatClassFactory.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public static class SeatClassFactory
+    {
+        public static ISeatClass FromString(string seatClass)
```
make this an extension method instead.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/DTOs/SeatClassFactory.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace AirportTicketBookingSystem.DTOs
+{
+    public static class SeatClassFactory
```
This is not a DTO, move it please to the ```Utilities``` folder.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/FlightService.cs:
```diff
@@ -0,0 +1,127 @@
+            }
+        }
+
+        public void ImportFlights(List<Flight> flights)
```
I think the import feature in our requirements is excepting a csv file as an input to import to our system

RandaZraik replied:
also we have requirements to validate imported data, and provide dynamic model validation details

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/CsvFilePathes.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public struct CsvFilePaths
+    {
+        private static readonly string BasePath = @"D:\fts-projects\CSVFiles\";
```
we shouldn't be hardcoding absolute paths, if I run this project on my machine, it will fail. You can use ```Path.Combine(AppDomain.CurrentDomain.BaseDirectory....``` or something similar.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Utilities/FlightBookingUtility.cs:
```diff
@@ -0,0 +1,45 @@
+
+namespace AirportTicketBookingSystem.Utilities
+{
+    public static class FlightBookingUtility
```
```FlightBookingUtilities```

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Utilities/PassengerBookingUtility.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace AirportTicketBookingSystem.Utilities
+{
+    public static class PassengerBookingUtility
```
PassengerBookingUtilities

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Utilities/FlightBookingUtility.cs:
```diff
@@ -0,0 +1,45 @@
+            }
+        }
+
+        public static void CascadeDeleteBookings(int flightId, List<Booking> bookings, Action<int> deleteBooking)
```
I think this is just overengineering, the only duplicate logic you are trying to unify here is the ```relatedBookings``` filtration, you are still passing a different deletion DB method. Removing ```CascadeDeleteBookings``` is better for simplicity. There's no business logic that you are trying to have in one place so it can only be changed once, or other similar scenarios that's worth the abstraction.

RandaZraik replied:
maybe you can just move deleteBooking method to the booking repo and call it from the other repos

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Utilities/FlightBookingUtility.cs:
```diff
@@ -0,0 +1,45 @@
+                deleteBooking(bookingId);
+            }
+        }
+        public static void CascadeAddBookings(Flight flight, Action<Booking> addBooking)
```
same comment as above

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Utilities/PassengerBookingUtility.cs:
```diff
@@ -0,0 +1,47 @@
+            }
+        }
+
+        public static void CascadeDeleteBookings(int passengerId, List<Booking> bookings, Action<int> deleteBooking)
```
same comment as above

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Utilities/PassengerBookingUtility.cs:
```diff
@@ -0,0 +1,47 @@
+            }
+        }
+
+        public static void CascadeAddBookings(Passenger passenger, Action<Booking> addBooking)
```
same comment as above

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/BookingRepo.cs:
```diff
@@ -0,0 +1,109 @@
+
+namespace AirportTicketBookingSystem.Repositories
+{
+    public class BookingRepo : IBookingRepo
```
```BookingRepository```. Same for the other repo and interfaces, this is more meaningful.

RandaZraik replied:
these four repos can be abstracted with a base class, something like:
```csharp
public interface IRepository<T> where T : class, IEntity
{
    Task<T> GetByIdAsync(int id);
    Task<List<T>> GetAllAsync();
    Task AddAsync(T entity);
    Task UpdateAsync(Func<T, bool> predicate, T entity);
    Task DeleteAsync(int id);
}

public abstract class CsvRepositoryBase<TEntity, TCsvModel> : IRepository<TEntity>
    where TEntity : class, IEntity
    where TCsvModel : class
{
}
```

I see there's  a different in the booking repo, different mapping, you can add to the abstract class some abstract methods for mapping models from/to csv  that can have different implementation per repo.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/BookingRepo.cs:
```diff
@@ -0,0 +1,109 @@
+            Save(bookingRecords);
+        }
+
+        public void UpdateBooking(Func<Booking, bool> predicate, Booking newBooking)
```
I don't see there is a need to pass a func, not sure why  we need this complication

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/FlightService.cs:
```diff
@@ -0,0 +1,127 @@
+                throw new KeyNotFoundException($"Flight with Id {flightId} does not exist");
+
+            var allBookings = _bookingRepo.GetAllBookings();
+            FlightBookingUtility.CascadeDeleteBookings(flightId, allBookings, _bookingRepo.DeleteBooking);
```
instead of these two lines, just call a repo method in bookingrepo that deletes bookings by flightid

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/ManagerService.cs:
```diff
@@ -0,0 +1,89 @@
+        {
+            var allBookings = _bookingRepo.GetAllBookings();
+
+            var query =
```
filtration should happen on the DB layer not logic/service layer, think about it as you are dealing with a real DB, and you are retrieving all bookings from db then doing the filtration, lets move the filtration logic (including getallbookings) to a method in the booking repo

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/PassengerService.cs:
```diff
@@ -0,0 +1,163 @@
+
+            FlightBookingUtility.AttachBookings(allFlights, allBookings);
+
+            var query =
```
same comment as above, filtration should happen in the repo

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/BookingRepo.cs:
```diff
@@ -0,0 +1,109 @@
+            )).ToList();
+        }
+
+        public void AddBooking(Booking booking)
```
when adding new booking, you wouldn't have an id for it yet, ids usually gets generated on the DB level, so you need to generate a unique one prior to save

RandaZraik replied:
same comment for the other add methods in the other repos

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/FlightRepo.cs:
```diff
@@ -0,0 +1,91 @@
+            var selectedFlight = flightRecords.FirstOrDefault(b => b.Id == id);
+
+            return selectedFlight ?? throw new KeyNotFoundException($"Flight with Id {id} was not found");
+            
```
remove extra newline

yazan-hamamdi committed:
Optimize file names and place classes and methods in the correct folders

yazan-hamamdi committed:
Resolved the over-engineering issue in Cascade and converted them into simple methods in their proper place

yazan-hamamdi committed:
Simplified update methods across all repositories

yazan-hamamdi committed:
Add an automatic ID generator

yazan-hamamdi committed:
Add batch flight upload and validation
Import flights from CSV file
Validate imported data and return detailed error list

yazan-hamamdi committed:
Add dynamic model validation details
Generate validation constraints dynamically for each flight data field

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Interfaces/IEntity.cs:
```diff
@@ -0,0 +1,8 @@
+﻿namespace AirportTicketBookingSystem.Interfaces
+{
+    public interface IEntity
```
this is more of a model, lets move to the models folder

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/FlightRepository.cs:
```diff
@@ -0,0 +1,175 @@
+            csv.WriteRecords(flights);
+        }
+
+        public List<Flight> SearchAvailableFlights(string departureCountry = null, string destinationCountry = null,
```
you would need a service method to call this repo, if you have an api to consume this method, you won't call the repo directly from the API layer, you would do it through a service. Lets do it with any other similar methods.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/BookingService.cs:
```diff
@@ -0,0 +1,95 @@
+            _bookingRepo.DeleteBooking(bookingId);
+        }
+
+        public void ModifyBooking(int passengerId, int bookingId, Booking newBooking)
```
NIT: `UpdateBooking` is more convenient

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Utilities/ToSeatClass.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace AirportTicketBookingSystem.Utilities
+{
+    public static class SeatClassExtensions
```
the file name should be `SeatClassExtensions` not `ToSeatClass`. Should match the class name.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/BookingRepository.cs:
```diff
@@ -0,0 +1,149 @@
+            return query.ToList();
+        }
+
+        public void DeleteBookings(Func<Booking, bool> predicate)
```
I don't think there's a need to pass a predicate here, an id is more straightfoward.

yazan-hamamdi replied:
`public void DeleteFlightWithBookings(int flightId)
{
    var flight = _flightRepository.GetFlightById(flightId);
    if (flight == null)
        throw new KeyNotFoundException($"Flight with Id {flightId} does not exist");

    _bookingRepository.DeleteBookings(b => b.FlightId == flightId);
    _flightRepository.DeleteFlight(flightId);
}`
`public void DeletePassengerWithBookings(int passengerId)
{
    var passenger = _passengerRepository.GetPassengerById(passengerId);
    if (passenger == null)
        throw new KeyNotFoundException($"Passenger with Id {passengerId} does not exist");

    _bookingRepository.DeleteBookings(b => b.PassengerId == passengerId);
    _passengerRepository.DeletePassenger(passengerId);
}`
I used a predicate here to keep the method flexible. In our case, we sometimes need to delete multiple bookings related to a flight or a passenger (foreign keys) not just by the bookings primary key

RandaZraik replied:
I see, but this is confusing to the user, if i see a method called Delete, i would just expect to pass an id to delete the entity.  Having a pretty generic method is not beneficial always, For your case and for better readability I would create another method called `DeleteByPassengerId`. You can have this type of abstraction but in a lower layer, and have a wrapper that uses this method that has a more meaningful name like `DeleteByPassengerId` that others can use. But I think for your case, just having a new method is good enough.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/FlightRepository.cs:
```diff
@@ -0,0 +1,175 @@
+            return query.ToList();
+        }
+
+        public List<string> ImportFlightsFromCsv(string csvFilePath)
```
i would move this to the service layer, and from there i call the db to save the records. Repos really just for doing straightforward db operations, any business logic should live in the service/domain layer

RandaZraik replied:
treat this as an external importer from csv files, i know we have db saved in csv files, but later it will be an actual db and you will need to import things to it from different resources. You can abstract csv main operations into a helper that can be re-used from repos and services.

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/FlightRepository.cs:
```diff
@@ -0,0 +1,175 @@
+            return errors;
+        }
+
+        public bool IsDuplicateFlight(Flight flight, List<Flight> existingFlights)
```
same thing, this method is not using the db storage at all, can go to the service layer

yazan-hamamdi committed:
Improved file organization and naming, and also placed methods in their proper locations

yazan-hamamdi committed:
Made the repositories generic

yazan-hamamdi committed:
Make the services generic

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Interfaces/IFlightService.cs:
```diff
@@ -0,0 +1,27 @@
+        TravelClass? seatClass = null,
+        decimal? minPrice = null,
+        decimal? maxPrice = null
+         );
```
Better indentation. Same for other methods.

```suggestion
        List<FieldValidationDetail> GetFlightModelValidationDetails();
            List<Flight> SearchAvailableFlights(
            string departureCountry = null,
            string destinationCountry = null,
            string departureAirport = null,
            string arrivalAirport = null,
            DateTime? departureDateFrom = null,
            DateTime? departureDateTo = null,
            TravelClass? seatClass = null,
            decimal? minPrice = null,
            decimal? maxPrice = null
         );
```

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/BaseService.cs:
```diff
@@ -0,0 +1,68 @@
+
+namespace AirportTicketBookingSystem.Services
+{
+    public abstract class BaseService<T> : IService<T> where T : class, IEntity
```
This is nice :D later on when working on more complicated services, you will notice that this might not be necessary if we have other logic to apply along side with CRUD. CRUD business logic can be different from one entity to another, and one entity for example Order would require 3 repos to create it (Order repo, order item repo, order logs), so you always end up not using these base methods here and overriding them.

Another note to keep in mind is that abstraction hides lots of things, we need to make sure we don't overengineer and fill in a trap where we have too many dependencies hidden.

This is just an FYI :D Nice work so far!

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/FlightService.cs:
```diff
@@ -0,0 +1,161 @@
+            return flight;
+        }
+
+        public void DeleteFlightWithBookings(int flightId)
```
if you are not going to use the delete method in `BaseService`, then we shouldn't introduce a new method, but override the existing one: `public override void Delete()`

RandaZraik replied:
same comment for the other methods in these services. If you end up not really using the methods in BaseService, you can get rid of that

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/FlightRepository.cs:
```diff
@@ -0,0 +1,31 @@
+    {
+        public FlightRepository(string filePath) : base(filePath) { }
+
+        public List<Flight> GetAllFromFile(string filePath)
```
the code you have here is not really repo related, just a utility to read from a csv file, should belong a utility class. I know we have the same logic in the base repo, you can update the base repo to use this utility as well

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/BookingService.cs:
```diff
@@ -0,0 +1,62 @@
+                throw new KeyNotFoundException($"Passenger with Id {passengerId} does not exist");
+
+            var allBookings = _bookingRepository.GetAll();
+            return allBookings.Where(b => b.PassengerId == passengerId).ToList();
```
filtration should happen on the db level, you shouldn't get all bookings here and do then the filtration. Add a new repo method to get booking by passenger id

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/FlightService.cs:
```diff
@@ -0,0 +1,161 @@
+            }
+        }
+
+        public List<FieldValidationDetail> GetFlightModelValidationDetails()
+        {
+            var details = new List<FieldValidationDetail>();
+            var properties = typeof(Flight).GetProperties();
+
+            foreach (var prop in properties)
+            {
+                var fieldDetail = new FieldValidationDetail
+                {
+                    FieldName = prop.Name,
+                    FieldType = prop.PropertyType.Name,
+                    Constraints = ValidationMetadataHelper.GetPropertyConstraints(prop)
+                };
+                details.Add(fieldDetail);
+            }
+            return details;
+        }
```
NIT:

```suggestion
        public List<FieldValidationDetail> GetFlightModelValidationDetails()
        {
            return typeof(Flight)
                .GetProperties()
                .Select(prop => new FieldValidationDetail
                {
                    FieldName = prop.Name,
                    FieldType = prop.PropertyType.Name,
                    Constraints = ValidationMetadataHelper.GetPropertyConstraints(prop)
                })
                .ToList();
        }
```

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/FlightService.cs:
```diff
@@ -0,0 +1,161 @@
+            FlightBookingUtilities.AttachBookings(allFlights, allBookings);
+
+            var query =
+            from f in allFlights
```
increase indentation for this line and rest of query

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/FlightService.cs:
```diff
@@ -0,0 +1,161 @@
+            );
+        }
+
+        private List<string> ValidateFlightForImport(Flight flight, List<Flight> existingFlights)
```
NIT:

```csharp
private List<string> ValidateFlightForImport(Flight flight, List<Flight> existingFlights)
{
    var flightIdentifier = $"{flight.DepartureCountry} -> {flight.DestinationCountry} on {flight.DepartureDate}";
    
    var duplicateError = IsDuplicateFlight(flight, existingFlights) 
        ? new[] { $"Duplicate flight: {flightIdentifier}" }
        : Enumerable.Empty<string>();
    
    var validationErrors = ValidationHelper.ValidateModel(flight)
        .Select(err => $"Flight {flightIdentifier}: {err}");
    
    return duplicateError.Concat(validationErrors).ToList();
}
```

yazan-hamamdi committed:
Improved SearchAvailableFlights query readability
Add CsvFileHelper
Improve the design and remove methods that I don't need to use in the service

yazan-hamamdi committed:
doing inner queries for search method

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Services/BookingService.cs:
```diff
@@ -0,0 +1,62 @@
+            if (booking == null || booking.PassengerId != passengerId)
+                throw new KeyNotFoundException("Booking not found for this passenger");
+
+            _bookingRepository.Delete(bookingId);
```
can't we just call the base delete method?

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/FlightRepository.cs:
```diff
@@ -0,0 +1,57 @@
+
+            var query =
+                from f in allFlights
+                join b in allBookings on f.Id equals b.FlightId into flightBookings
```
thumbs up

RandaZraik commented on AirportTicketBookingSystem/AirportTicketBookingSystem/Repositories/BookingRepository.cs:
```diff
@@ -0,0 +1,60 @@
+            if (!File.Exists(_filePath))
+                throw new FileNotFoundException("CSV file not found", _filePath);
+
+            using var reader = new StreamReader(_filePath);
```
you can add to the base repo the following to call here. 

```csharp
public virtual List<TModel> GetAll<TModel>() where TModel : class, new()
{
    return CsvFileHelper.ReadFromCsv<TModel>(_filePath).ToList();
}
```

RandaZraik replied:
FYI, the better way is to update base repo to actually have two models:

```csharp
public abstract class BaseRepository<TDomain, TDb> : IRepository<TDomain> 
    where TDomain : class, IEntity
    where TDb : class, IEntity
```

Let's skip this for now, this is just an FYI. Later on, when we cover mappers and introduce layered models (database and domain), this approach will make more sense.

yazan-hamamdi committed:
Refactored BookingRepository.GetAll to reuse BaseRepository.GetAll<T>

yazan-hamamdi committed:
Move passenger bookings filtration to repository level by adding GetByPassengerId
